# Lab Exercises

1) Let random variable $X$ be the number we get when we roll a die,    
    1. Write out the probability mass function of X.     
    2. Calculate the mean of X.  
    3. Calculate the variance of X.  

**Probabiliy mass function**

|x        |  1  |  2  |  3  |  4  |  5  |  6  |
|:--------|:---:|:---:|:---:|:---:|:---:|:---:|
|P(X = x) | 1/6 | 1/6 |1/6  | 1/6 |1/6  | 1/6 |



**mean of X**
E(X) = 1$*$P(X=1) + 2$*$P(X=2) + 3$*$P(X=3) + 4 $*$ P(X=4) + 5$*$ P(X=5) + 6$*$P(X=6)

E(X) = 3.5

**variance of X**
E^2(X) = (1-3.5)^2$*$P(X=1) + (2-3.5)^2$*$P(X=2) + (3-3.5)^2 $*$ P(X=3) + (4-3.5)^2$*$ P(X=4) + (5-3.5)^2$*$ P(X=5) +(6-3.5)^2$*$P(X=6)

E^2(X) = 2.92

2) Use either `randint` or `random_sample` to generate a series of 20 rolls. Use `.mean()` and `.var()` to calculate the sample mean and variance. How do these compare to the population mean and variance you calculated above? What if you increase the number of rolls to 2000. Are your estimates closer to the true statistics or further away?

In [1]:
import numpy as np


In [68]:
def roll_dice(number_of_rolls):
    """roll a die this many of times"""
    
    rolls = [np.random.randint(1,7) for _ in range(number_of_rolls)]
    
    mean = np.mean(rolls)
    variance = np.var(rolls)
    
    print("The mean of {} rolls is {}:".format(number_of_rolls,mean)) 
    print("The variance of {} rolls is {}:".format(number_of_rolls,variance))
    
    #return rolls

In [70]:
roll_dice(20)
roll_dice(2000)
#As you increase the number of rolls, your estimates move closer to the true statistics value.

The mean of 20 rolls is 4.1:
The variance of 20 rolls is 3.69:
The mean of 2000 rolls is 3.5105:
The variance of 2000 rolls is 2.9248897499999997:


3) Now we have two dice, and X and Y are the numbers appear on each die respectively. Let Z be the sum of the numbers from the two dice,  
    1. Write out the probability mass function of Z.  
    2. Calculate the mean of Z.  
    3. Calculate the variance of Z.

# Question Three


**Probabiliy mass function**

|x         |  2  |  3  |  4  |  5     |  6  | 7   |  8  |  9  |  10  |  11  |  12  |
|:--------|:---:|:---:|:---: |:---:   |:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|P(Z = z) | 1/36 |2/36  | 3/36 | 4/36 | 5/36 |6/36|5/36 | 4/36|3/36 |2 / 36| 1/36 |


**mean of Z**

E(Z) = 2$*$(1/36)+3$*$(2/36)+4$*$(3/36)+5$*$(4/36)+6$*$(5/36)+7$*$(6/36)8$*$(5/36)+9$*$(4/26)+10$*$(3/36)+11$*$(2/36)+12$*$ (1/36)

E(Z) =  7.0

**variance of Z**

(2-7)$*$(1/36)+ (3-7)$*$(2/36)+(4-7)$*$(3/36)+ (5-7)$*$(4/36)+(6-7)$*$(5/36)+(7-7)$*$(6/36)+(8-7)$*$(5/36) +(9-7)$*$(4/26)+(10-7)$*$(3/36) +(11-7)$*$(2/36)+(12-7)$*$ (1/36)

Var(Z) = 5.83


4) Do the same as in (2) but with two dice as in (3). Again, check to see how $N$, the number of samples, affects your result.

In [54]:
def roll_two_dice(number_of_times):
    """roll two dice. Roll each one individualy and then add each roll together"""
    
    dice_one = [np.random.randint(1,7) for _ in range(number_of_times)]
    dice_two = [np.random.randint(1,7) for _ in range(number_of_times)]
    
    combined_roll=[(x+y) for (x,y) in zip(dice_one,dice_two)]
    
    print('Your mean is {} for {} rolls'.format(np.mean(combined_roll),number_of_times))
    print('Your variance is {} for {} rolls'.format(np.var(combined_roll),number_of_times))
        


In [75]:
roll_two_dice(20)
roll_two_dice(2000)
#As you increase the number of rolls, your statistics move closer to the true population value

Your mean is 7.55 for 20 rolls
Your variance is 5.7475000000000005 for 20 rolls
Your mean is 6.987 for 2000 rolls
Your variance is 6.016831 for 2000 rolls


5) Find the variance of each of the following bets from the class notes. Which bet is riskiest and which best is safest (where "risk" is assessed according to variance, not expected value)?

1. You get \$5 with probability 1.0.  
2. You get \$10 with probability 0.5, or \$0 with probability 0.5.  
3. You get \$5 with probability 0.5, \$10 with probability 0.25 and \$0 with probability 0.25.  
4. You get \$5 with probability 0.5, \$105 with probability 0.25 or lose \$95 with probability 0.25.  

# Question



6) Brandon Lang is a creative entrepreneur who has developed a novelty soap item called Jackpot to target consumers with a gambling habit. Inside each bar of Jackpot shower soap is a single rolled-up bill of U.S. currency. The currency (rolled up and sealed in shrink-wrap) is appropriately inserted into the soap mixture prior to the cutting and stamping procedure. The distribution of paper currency (per 1000 bars of soap) is given in the following table.  

|   Bill Denomination   |   Number of Bills |
|:---------------------:|:-----------------:|
|\$1                     | 520 |
|\$5                     | 260 |
|\$10                    | 120 |
|\$20                    | 70  |
|\$50                    | 29  |  
|\$100                   | 1   |
| Total                  | 1000|

1. What is the expected amount of money in a single bar of Jackpot soap?
2. What is the standard deviation of the money in a single bar of Jackpot soap?
3. How many bars of soap would a customer have to buy so that, on average, he or she has purchased three bars   containing a \$50 or \$100 bill?
4. If a customer buys 8 bars of soap, what is the probability that at least one of these bars contains a bill of \$20 or larger? 

7) Write a function that randomly draws from the above distribution. 

1. Use it to simulate the answer to part 3 of the last question at least 20 times. How often do you get the expected result? 
2. Use it to simulate part 4 of the last question 100 times. How close does your simulation come to the true expected value?

In [82]:
import numpy as np
import itertools 

In [416]:
def distribution(n):
    numbers =[np.random.choice([1,5,10,20,50,100],p=[.52,.26,.12,.07,.029,.001]) for _ in range(n)]
    
    if len(numbers) ==1:
        for item in numbers:
            return(int(item))
    else:
        return numbers
    
def pick_value (number_of_bars,*args):
    """Will return the number of purchases it took to have a number of bars of your desired value"""
    
    
    value_of_bars = args
    count_of_draws = 0
    count_of_bars_with_value_we_want = 0
    
    while count_of_bars_with_value_we_want<number_of_bars:  ##keep purchasing bars until you reach the number you want
        choice = distribution(1)
        #print(type(choice))
        #print(choice, ' this is your choice')
   
        if choice == value_of_bars[0] or choice ==value_of_bars[1]:
            #print('choice should equal item')
            count_of_draws+=1
            count_of_bars_with_value_we_want +=1
        else:
            count_of_draws+=1
                
    

    #print(count_of_draws,'number of bar purchases')
    #print(count_of_bars_with_value_we_want, ' bars we want with denomination {}'.format(value_of_bars))
    return count_of_draws
    

In [417]:
pick_value(3,50,100)

110

In [420]:
simulate_q_3 =[pick_value(3,50,100) for item in range(20)]

print(np.mean(simulate_q_3))

122.05


# Question seven

The mean should be the expected value of the list of picking three bars of soap with values 50  or  100. The mean should center around the expected value of 100 (which it does).

In [410]:
#If a customer buys 8 bars of soap,
#what is the probability that at least one of these bars contains a bill of $20 or larger?

def twenty_or_larger(purchases):
    """how many purchases will have bars with values of 20 or more"""
    numbers =[np.random.choice([1,5,10,20,50,100],p=[.52,.26,.12,.07,.029,.001]) for _ in range(purchases)]
    count = 0
    for item in numbers:
        if item >=20:
            count +=1
    return count/purchases

In [411]:
twenty_or_larger(8)

0.125

# Question seven part two

The probability that you receive at least on soap bar of $20 or more is 12.5%

# 8)
(Extra Credit) Think back to the coin tossing example we discussed at the beginning of the lecture, if we let $X$ be the number of heads we get from 10 coin tosses, what is the probability mass function of $X$? What are the mean and variance of $X$? (**Note**: You are not required to complete this problem, so don't spend too much time on this. This is just to get you to start thinking about the probabilities of events when we have multiple trials of an experiment - we will talk about this on Monday.)

The binomial distribution is n! / ( (n-x)!$*$x!) $*$ p ^x  q ^(n-x)

PDF

X=1  -> 10! /$ ((10-1)!*1!)*.5^1*.5^9$

X=2 -> 10! /$ ((10-2)!*2!)*.5^2*.5^8$

X=3  -> 10! /$ ((10-3)!*3!)*.5^3*.5^7$

X=4 -> 10! /$ ((10-4)!*4!)*.5^4*.5^6$

X=5 -> 10! /$ ((10-5)!*5!)*.5^5*.5^5$

X=6 -> 10! /$ ((10-6)!*6!)*.5^6*.5^4$

X=7-> 10! /$ ((10-7)!*7!)*.5^7*.5^3$

X=8-> 10! /$ ((10-8)!*8!)*.5^8*.5^2$

X=9-> 10! /$ ((10-9)!*9!)*.5^9*.5^1$

X=10 -> 10! /$ ((10-10)!*9!)*.5^9*.5^1$
 